### ISSUE TITLE
Image pull using secrets

### ISSUE THREAD

**Issue by pmorie** [GoogleCloudPlatform/kubernetes#7627]
There are a couple in-progress items currently that are going to touch the image pull with secret story, so I'm creating this issue to get everyone on the same page and hopefully do some productive design and factorings in front of the implementation work that will need to happen.

Our long-term goal is to be able to use secrets to enable pull from private image repositories.  Those secrets will definitely come from service accounts; it seems possible to me that we'll want to enable a use-case where a secret can also be explicitly referenced in a pod's spec to enable image pull as well.

One key thing to figure out is exactly how we want to model this inside the kubelet.  Currently the container manager has a `Pull` method that simply takes the container's `image` field as input.  I think, regardless of how the kubelet ultimately models associations between credentials, pods, and images, that this signature will need to change to include:

1.  Information about the pod
2.  The credentials to be used for the pull

Some random questions I can think of that we will need to sort out:

1.  Say that two different pods are set up to pull that same image from the same private registry using different credentials.  If the kubelet pulls the image on behalf of pod X, and then pod Y is scheduled onto the same node, and the image is already present on the node, do we still have to test pod Y's credentials somehow?  I would say the answer is yes, but I don't believe there's a way to dry-test a set of credentials with a docker registry at this point.  Please correct me if I'm wrong -- I would love to be wrong about this.
2.  Presumably the problem in (1) applies to each auth mode, credential type, registry type.
3.  Will credentials be cached on the node?  If so we need to have tmpfs storage for these, since they shouldn't be allowed to come to rest on the node's fs.

@liggitt @yifan-gu @smarterclayton @vmarmol 

Trying to keep the tagging minimal, feel free to tag your friends and coworkers.

**Comment by smarterclayton** [comment]
----- Original Message -----
> There are a couple in-progress items currently that are going to touch the
> image pull with secret story, so I'm creating this issue to get everyone on
> the same page and hopefully do some productive design and factorings in
> front of the implementation work that will need to happen.
> 
> Our long-term goal is to be able to use secrets to enable pull from private
> image repositories.  Those secrets will definitely come from service
> accounts; it seems possible to me that we'll want to enable a use-case where
> a secret can also be explicitly referenced in a pod's spec to enable image
> pull as well.
> 
> One key thing to figure out is exactly how we want to model this inside the
> kubelet.  Currently the container manager has a `Pull` method that simply
> takes the container's `image` field as input.  I think, regardless of how
> the kubelet ultimately models associations between credentials, pods, and
> images, that this signature will need to change to include:
> 
> 1.  Information about the pod
> 2.  The credentials to be used for the pull
> 
> Some random questions I can think of that we will need to sort out:
> 
> 1.  Say that two different pods are set up to pull that same image from the
> same private registry using different credentials.  If the kubelet pulls the
> image on behalf of pod X, and then pod Y is scheduled onto the same node,
> and the image is already present on the node, do we still have to test pod
> Y's credentials somehow?  I would say the answer is yes, but I don't believe
> there's a way to dry-test a set of credentials with a docker registry at
> this point.  Please correct me if I'm wrong -- I would love to be wrong
> about this.

In the long run yes, which requires that we have a way in the container to check credentials without initiating a pull.  In the medium term, we just initiate a pull.  In the very short term, we do nothing.

> 2.  Presumably the problem in (1) applies to each auth mode, credential type,
> registry type.
> 3.  Will credentials be cached on the node?  If so we need to have tmpfs
> storage for these, since they shouldn't be allowed to come to rest on the
> node's fs.

No different than secrets I don't think.

**Comment by liggitt** [comment]
@deads2k fyi

**Comment by roberthbailey** [comment]
@dchen1107 @erictune

**Comment by deads2k** [comment]
A couple questions come to mind.
 1. Should there be a merge order for pod/service account pull credentials, gcp credentials, and `~/.dockercfg` credentials?  I think I'm against this, but it seems reasonable to argue.  I'd say that if the pod and/or the relevant  service account come with pull credentials (let's call them dynamic creds), those are the only ones we should use.  If they do not, then the static creds (gcp and `~/.dockercfg`) should be used instead.  That allows consistent behavior across nodes if you take the time to set up your service account, but it doesn't force someone to make that work out of the box.

 1. Should we attempt multiple credentials for a single pull?  Let's say we go with the "either static or dynamic credentials, not both" approach.  We still have to deal with the case where we have colliding host and path options.  The current implementation is last one wins in a non-deterministic ordering (probably a bug): https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/credentialprovider/keyring.go#L51 .  I would claim that if there are multiple matching entries inside of the credentials we've decided to use, we attempt each them in their order of specificity.  This allows for rotation.

**Comment by smarterclayton** [comment]
----- Original Message -----
> A couple questions come to mind.
>  1. Should there be a merge order for pod/service account pull credentials,
>  gcp credentials, and `~/.dockercfg` credentials?  I think I'm against this,
>  but it seems reasonable to argue.  I'd say that if the pod and/or the
>  relevant  service account come with pull credentials (let's call them
>  dynamic creds), those are the only ones we should use.  If they do not,
>  then the static creds (gcp and `~/.dockercfg`) should be used instead.
>  That allows consistent behavior across nodes if you take the time to set
>  up your service account, but it doesn't force someone to make that work out
>  of the box.

I think that's reasonable.  Explicit secrets "use these secrets to get images" trumps "grab any old random thing on the host".

> 
>  1. Should we attempt multiple credentials for a single pull?  Let's say we
>  go with the "either static or dynamic credentials, not both" approach.  We
>  still have to deal with the case where we have colliding host and path
>  options.  The current implementation is last one wins in a
>  non-deterministic ordering (probably a bug):
>  https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/credentialprovider/keyring.go#L51
>  .  I would claim that if there are multiple matching entries inside of the
>  credentials we've decided to use, we attempt each them in their order of
>  specificity.  This allows for rotation.

Rotation will be important.  If we make the order deterministic with most specific / ordered, that also makes sense to me.

**Comment by pmorie** [comment]
I suggest the following changes to the container runtime interface:

```go
// ImageSpec is a kubelet-internal API to describe images.  At first
// it will be just a wrapper around the API's string-typed image field
// but eventually we will add more detailed APIs for different image
// formats.
type ImageSpec struct {
  Image string
}

type Runtime interface {
  // other methods omitted
  PullImage(spec ImageSpec, secrets []api.Secret) error
  IsImagePresent(spec ImageSpec) (bool, error)
  RemoveImage(spec ImageSpec) error
}
```

The `ImageSpec` type is to design-ahead for #7203 and eventual support for arbitrary image formats.

The container runtime will be injected with a list of possible secrets (eventually coming from the pod's service account) to use to perform the image pull.  This way, the container runtime can make its own decisions about how to use secrets to perform pulls, and the kubelet doesn't need to be aware of strategies for how the secrets are used.

I do agree that if secrets are supplied, you should not be able to fail back to the kubelet's default dockercfg, and that should be part of the stated contract for the `PullImage` method in the interface.

If we agree on this API, I will make a pull with the above changes today that just injects an empty slice for secrets.

**Comment by pmorie** [comment]
@vmarmol @deads2k @smarterclayton @yifan-gu thoughts? (cc @liggitt)

**Comment by yifan-gu** [comment]
@pmorie Thank you for pinging me! Will take a look!

/cc @jonboulle @philips

**Comment by bgrant0607** [comment]
I would prefer to have a secret-based mechanism than to make it easier to roll out a cluster-wide dockercfg (https://github.com/GoogleCloudPlatform/kubernetes/issues/1627#issuecomment-69143858, #2603), though perhaps we need both.

**Comment by pmorie** [comment]
@bgrant0607, I think we need both.

**Comment by bgrant0607** [comment]
FWIW, discussion on #7150 may be relevant, or may not, since I didn't actually read it. :-)

**Comment by pmorie** [comment]
Noted, lol

**Comment by pmorie** [comment]
Resolved by #8582 #7974, among others.
